google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryWriteManualClientTest.java

```diff
@@ -406,6 +409,208 @@ public void testJsonStreamWriterWithDefaultStream()
+        assertEquals("ddd", temp.get(1).getStringValue());
+      }
+      assertFalse(rowsIter.hasNext());
+    }
+  }
+
+  @Test
+  public void testJsonStreamWriterSchemaUpdateConcurrent()
+      throws DescriptorValidationException, IOException, InterruptedException {
+    // Create test table and test stream
+    String tableName = "ConcurrentSchemaUpdateTestTable";
+    TableId tableId = TableId.of(DATASET, tableName);
+    Field col1 = Field.newBuilder("col1", StandardSQLTypeName.STRING).build();
+    Schema originalSchema = Schema.of(col1);
+    TableInfo tableInfo =
+        TableInfo.newBuilder(tableId, StandardTableDefinition.of(originalSchema)).build();
+    bigquery.create(tableInfo);
+    TableName parent = TableName.of(ServiceOptions.getDefaultProjectId(), DATASET, tableName);
+    WriteStream writeStream =
+        client.createWriteStream(
+            CreateWriteStreamRequest.newBuilder()
+                .setParent(parent.toString())
+                .setWriteStream(
+                    WriteStream.newBuilder().setType(WriteStream.Type.COMMITTED).build())
+                .build());
+
+    // Create test JSON objects
+    JSONObject foo = new JSONObject();
+    foo.put("col1", "aaa");
+    JSONArray jsonArr = new JSONArray();
+    jsonArr.put(foo);
+
+    JSONObject foo2 = new JSONObject();
+    foo2.put("col1", "bbb");
+    JSONArray jsonArr2 = new JSONArray();
+    jsonArr2.put(foo2);
+
+    JSONObject updatedFoo = new JSONObject();
+    updatedFoo.put("col1", "ccc");
+    updatedFoo.put("col2", "ddd");
+    JSONArray updatedJsonArr = new JSONArray();
+    updatedJsonArr.put(updatedFoo);
+
+    // Prepare updated schema
+    Field col2 = Field.newBuilder("col2", StandardSQLTypeName.STRING).build();
+    Schema updatedSchema = Schema.of(ImmutableList.of(col1, col2));
+    TableInfo updatedTableInfo =
+        TableInfo.newBuilder(tableId, StandardTableDefinition.of(updatedSchema)).build();
+
+    // Start writing using the JsonWriter
+    try (JsonStreamWriter jsonStreamWriter =
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build()) {
+      int numberOfThreads = 2;
+      ExecutorService streamTaskExecutor = Executors.newFixedThreadPool(5);
+      CountDownLatch latch = new CountDownLatch(numberOfThreads);
+      // Used to verify data correctness
+      AtomicInteger next = new AtomicInteger();
+      for (int i = 0; i < numberOfThreads; i++) {
+        streamTaskExecutor.submit(
+            () -> {
+              // write 2 rows of aaa on each Thread
+              for (int j = 0; j < 2; j++) {
+                try {
+                  jsonStreamWriter.append(jsonArr);
+                  next.getAndIncrement();
+                } catch (IOException | DescriptorValidationException e) {
+                  e.printStackTrace();
+                }
+              }
+
+              // update TableSchema
+              Table updatedTable = bigquery.update(updatedTableInfo);
```

Comment:
I think you should only issue one update request to the table, so move this out and execute it on another thread.

Reply:
also increase the thread count of append.

Reply:
sure - updating code to update table in another Thread.

Reply:
Increased Thread count of append from 2 to 5.